clang-tidy: use auto

neheb opened the pull request:
Found with modernize-use-auto

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
clang-tidy: use auto

Found with modernize-use-auto

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
manual std::accumulate + for loop conversions

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
clang-tidy: check for empty instead of size

Found with readability-container-size-empty

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
clang-tidy: don't empty initialize std::string

Found with readability-redundant-string-init

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
clang-tidy: use uppercase number literals

Found with readability-uppercase-literal-suffix

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
clang-tidy: simplify boolean expressions

Found with readability-simplify-boolean-expr

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
clang-tidy: remove pointless returns

Found with readability-redundant-control-flow

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
clang-tidy: fix inconsistent declarations

Found with readability-inconsistent-declaration-parameter-name

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
clang-tidy: don't check for null pointer with delete

Found with readability-delete-null-pointer

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
clang-tidy: remove static in anonymous namespace

Found with readability-static-definition-in-anonymous-namespace

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
clang-tidy: don't use compare

Found with readability-string-compare

clang-tidy: check for empty instead of empty string

Found with readability-container-size-empty

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
clang-tidy: use const reference

Found with performance-for-range-copy

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
clang-tidy: access static through instance

Found with readability-static-accessed-through-instance

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
clang-tidy: fix if identation problems

Found with readability-misleading-indentation

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
clang-tidy: change C includes to C++ ones

Found with modernize-deprecated-headers

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
clang-tidy: remove pointless member inits

Found with modernize-use-default-member-init

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
clang-tidy: remove pointless public

Found with readability-redundant-access-specifiers

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
clang-tidy: convert several member functions to const

Found with readability-make-member-function-const

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
manual algorithm conversions

These were supposed to be commited but got lost in a rebasing error.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
clang-tidy: manual using conversions

clang-tidy doesn't want to automatically fix these.

Found with modernize-use-using.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

piponazo commented:
Hi @neheb , thanks again for this wonderful contribution! :clap: 

I will start to review this PR soon. I just wanted to let you know that the windows builds are failing at the moment with this error:

```
..\src\FileIo.cpp(302): error C3861: 'WError': identifier not found
```

Let me know if you need help to debug these problems on Windows.

neheb commented:
Haha what a typo. Werror vs. WError. I'll fix as soon as I can.

piponazo commented on src/preview.cpp:
```diff
@@ -939,9 +938,8 @@ namespace {
-        unsigned long decodeBase64Table[256] = {};
-        for (unsigned long &i : decodeBase64Table)
-            i = invalid;
+        unsigned long decodeBase64Table[256];
```
This is a good candidate to be transformed to `std::array`. Then we could also use `decodeBase64Table.fill(...)`

neheb replied:
I tried that. Didn't want to compile under Windows.

piponazo reviewed (changes requested):
In general it looks great! I really appreciate your efforts in modernising the code base. 

The only changes I would drop are the `readability-avoid-const-params-in-decls` ones.

And we also need to verify the changes in `types.hpp` which could affect consumers of Exiv2.

neheb commented on samples/Jzon.cpp:
```diff
@@ -60,36 +60,34 @@ namespace Jzon
-			return spacing;
-		}
+            return std::string(format.indentSize * level, indentationChar);
+        }
```
I'm using that file yes.

Doing clang-format in one commit causes massive changes. Which is why I decided to do it gradually.

neheb commented on src/RemoteIo.cpp:
```diff
@@ -228,7 +228,7 @@ namespace Exiv2
                 throw Error(kerErrorMessage, "Data By Range is empty. Please check the permission.");
             }
-            const byte* source = reinterpret_cast<const byte*>(data.c_str());
+            const auto source = reinterpret_cast<const byte*>(data.c_str());
```
So the reason auto is used here is to avoid repeating the type name, which is already on the right. Original code was wrong. It should have been const byte *.

neheb replied:
Note that also this is different than the original. const auto evaluates to const byte* const, which is more correct I think. The source variable is not modified.

neheb commented on include/exiv2/types.hpp:
```diff
@@ -267,11 +267,11 @@ using Blob = std::vector<byte>;
         //@{
         //! Return iterator pointing to first element of buffer
-        byte* begin() noexcept;
+        byte* begin() const noexcept;
```
Removed.

neheb commented on src/sonymn_int.hpp:
```diff
@@ -77,8 +77,8 @@ namespace Exiv2 {
 
-    DataBuf sonyTagDecipher(uint16_t, const byte*, uint32_t, TiffComponent* const);
-    DataBuf sonyTagEncipher(uint16_t, const byte*, uint32_t, TiffComponent* const);
+    DataBuf sonyTagDecipher(uint16_t, const byte*, uint32_t, TiffComponent*);
```
Done.

neheb commented on include/exiv2/jpgimage.hpp:
```diff
@@ -82,15 +82,15 @@ namespace Exiv2
         ///  -2 if the pPsData buffer does not contain valid data.
         static int locateIrb(const byte* pPsData, long sizePsData, uint16_t psTag, const byte** record,
-                             uint32_t* const sizeHdr, uint32_t* const sizeData);
+                             uint32_t* sizeHdr, uint32_t* sizeData);
```
Removed.

neheb commented:
```
In file included from /usr/include/stdlib.h:958:0,

                 from /usr/include/c++/5/cstdlib:72,

                 from /home/travis/build/Exiv2/exiv2/samples/exiv2json.cpp:10:

In function ‘char* realpath(const char*, char*)’,

    inlined from ‘void fileSystemPush(const char*, Jzon::Node&)’ at /home/travis/build/Exiv2/exiv2/samples/exiv2json.cpp:235:31:

/usr/include/x86_64-linux-gnu/bits/stdlib.h:43:91: error: call to ‘__realpath_chk_warn’ declared with attribute warning: second argument of realpath must be either NULL or at least PATH_MAX bytes long buffer [-Werror]

cc1plus: all warnings being treated as errors
```

Impressive. This PR actually uncovered a bug.

neheb committed:
exiv2json: use PATH_MAX

Older versions of glibc mandate that realpath is called with PATH_MAX
or NULL.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb commented:
https://sourceware.org/git/?p=glibc.git;a=blob;f=stdlib/bits/stdlib.h;h=9134d3f36b5be5cb1b5a5874a622f8338b5d2288;hb=HEAD#l41

Cool.

neheb committed:
clang-tidy: don't use else after return

Found with readability-else-after-return

Signed-off-by: Rosen Penev <rosenp@gmail.com>
